Implement support for the RunCam protocol

andyp1per opened the pull request:
Implement support for the RunCam protocol.
```
CAM_RC_FEATURES - manually set RunCam features, but should be determined automatically
CAM_RC_BT_DELAY - time it takes for the runcam to boot
CAM_RC_BTN_DELAY - time to wait between button presses
CAM_RC_MDE_DELAY - time to wait between mode button presses
CAM_RC_TYPE - device type
CAM_RC_CONTROL - control access to OSD menu, bit 1 is yaw right, bit 2 is middle position of a 
3-button switch (rc option 78), bit 3 is a two-position switch (rc option 79). All can be enabled at the 
same time
```
In order to enable the support you need something like:
```
RC8_OPTION = 78
SERIALn_PROTOCOL = 26
```
You can also use option 79 to control OSD menu entry/exit rather than stick control.
This now all works:
* Video start/stop is now working off the switch defined by RC option 78
* The OSD menu can be entered by right yaw when throttle in in mid point and exited via left yaw when throttle in mid point
* Menus are navigated up/down via pitch and enter/exit/change via roll

It turns out I don't have a RunCam that implements the 5 key OSD protocol, so I cannot test that aspect. I have implemented the simpler 2 key OSD protocol. This requires firmware 2.4.4 on a split micro to work properly.

andyp1per commented:
Fixes https://github.com/ArduPilot/ardupilot/issues/10103

peterbarker commented:
Not following this PR - 'though it sounds awesome.

Is there any reason this is Copter-specific?

andyp1per commented:
@peterbarker no, other than the fact that I have a copter with a RunCam 🙂 
When I have it working I will make sure I add in support for other platforms as well.

peterbarker commented:
On Thu, 14 Nov 2019, Andy Piper wrote:

> @peterbarker no, other than the fact that I have a copter with a RunCam 🙂
> When I have it working I will make sure I add in support for other platforms as well.

So long as you're aware - particularly on things like the RC channel stuff 
- that code will need to be moved from (e.g.) ArduCopter/RC_Channel.cpp to 
libraries/RC_Channel/RC_Channel.cpp

andyp1per commented:
@peterbarker the BF code allows camera OSD control via RC commands when not armed. How do I access the main RC channels in a sensible way without depending on copter? (Copter has specific member variables for roll/pitch/yaw/throttle)

andyp1per commented:
Video start/stop is now working.

andyp1per commented:
I am successfully able to navigate OSD menus now, but it seems there is a bug in the protocol such that you can't actually change any settings this way. I have raised a support ticket with RunCam so we shall see. The split micro (what I have) does not support the 5-Key OSD simulation.

andyp1per commented:
I've bricked my runcam trying beta firmware. Development will be stopped for a while as I try and recover ☹️

andyp1per commented:
Incredibly, I have unbricked my runcam. This was really involved - I'll jot down the steps here in case anyone else has the same issues:

- Find a Windows 7 machine. I abjectly failed to get this to work on Windows 10.
- Install the Windows drivers only from https://support.runcam.com/hc/en-us/articles/360035371854-Flash-firmware-via-computer-for-Split-mini-2-if-it-is-bricked-during-firmware-updating- on Windows 7 this should work without having to do any messing with driver signing, but if you are *not prompted to install three device drivers* then something went wrong so you should follow the instructions in firmware-reflash-manual.pdf to disable driver signing
- Solder the USB port on your runcam, this involves wiring up the regular VCC and GND and then D- and D+ as shown here:
![image](https://user-images.githubusercontent.com/2893260/69899773-f9fc3680-1362-11ea-9e9b-e00c16300800.png)
Note that on my device at least the color coding was what is documented in the article above (D- white, D+ green) which is opposite to some articles on the internet.
- Connect the runcam to a *USB 2.0* port (USB 3.0 does not work)
- Wait for the 3 devices to be installed. On my device this took a long time and I only ended up with 2:
![image](https://user-images.githubusercontent.com/2893260/69899803-768f1500-1363-11ea-9e39-60dcc239c2e8.png)
I then had to manually install the spca6350 camera driver for the unrecognized device in device manager.
- Follow the rest of the instructions in firmware-reflash-manual.pdf but do *not* use the software in that download directory instead use the beta firmware package for the software: https://runcam.zendesk.com/attachments/token/Sq9AkN6bHCWTTKCG21y2ndQzv/?name=AQY006_V2.4.4_20191009.7z When I tried the older software I kept getting the error "ISP Core Error!"
- Even then I got a verify error but things seemed to be working and I was able to update to the regular firmware after that.

andyp1per commented:
And, Eureka! The beta runcam firmware 2.4.4 allows the settings to be changed. Pretty close to publishing this now.

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        return;
+    }
+
+    uint8_t *b = _buffer;
```
This is 64 bytes.  We can afford to put that on the stack.

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+
+    *b++ = RCDEVICE_PROTOCOL_HEADER;
+    *b++ = command;
+    if (param_len > 0) {
```
OK, this line confused things, really.  I think things might be clearer to have a `have_param` boolean...

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+}
+
+// handle a device info response
+void AP_RunCam::parse_device_info(RunCamResponse* response)
```
Take a const-reference here.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        // make sure the recording state reflects the control button
+        if (_video_recording) {
+            simulate_camera_button(RCDEVICE_PROTOCOL_CAM_CTRL_START_RECORDING);
+        }
```
whitespace

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        return false;
+    }
+
+    if (AP_HAL::millis() > _boot_delay_ms) {
```
Early-return

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+}
+
+// error handler for OSD simulation
+void AP_RunCam::simulation_OSD_cable_failed(RunCamResponse* response)
```
Take a const reference

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+    _user_info(userInfo),
+    _parser_func(parserFunc)
+{
+    _timeout_timestamp = AP_HAL::millis() + timeout;
```
Adding times is typically a mistake due to integer overflow issues.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+}
+
+// add a response to the response queue and return success/failure
+bool AP_RunCam::RunCamResponseQueue::push(AP_RunCam::RunCamResponse* response)
```
Do we really need a response queue rather than process everything synchronously?

If we do, then we have queue primitives already.  See `ObjectBuffer`.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    {
+        AP_Param::setup_object_defaults(this, var_info);
+        _rcmapper = rcmapper;
+        _singleton = this;
```
Need to add the fatal error if two instances created.

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+        AP_Param::setup_object_defaults(this, var_info);
+        _rcmapper = rcmapper;
+        _singleton = this;
+        _video_recording = true;
```
Use in-header initialisation rather than assigning in constructor.

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    }
+
+    // feature flag - uint16_t
+    enum RCFeatures
```
Use `enum class`.  Can be shortened to `Features`

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+        return _singleton;
+    }
+
+    // feature flag - uint16_t
```
Incorrect comment

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    // feature flag - uint16_t
+    enum RCFeatures
+    {
+        RCDEVICE_PROTOCOL_FEATURE_SIMULATE_POWER_BUTTON = (1 << 0),
```
```suggestion
        SIMULATE_POWER_BUTTON = (1 << 0),
```

etc

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    };
+
+    // operation of camera button simulation
+    enum RCCameraControlOperation
```
```suggestion
    enum class ControlOperation
```

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    // operation of camera button simulation
+    enum RCCameraControlOperation
+    {
+        RCDEVICE_PROTOCOL_CAM_CTRL_SIMULATE_WIFI_BTN = 0x00, // WiFi/Mode button
```
```suggestion
        SIMULATE_WIFI_BTN = 0x00, // WiFi/Mode button
```

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    };
+
+    // camera control commands
+    enum RCCameraCommands
```
```suggestion
    enum class Command
```

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    // camera control commands
+    enum RCCameraCommands
+    {
+        RCDEVICE_PROTOCOL_COMMAND_GET_DEVICE_INFO = 0x00,
```
```suggestion
        GET_DEVICE_INFO = 0x00,
```

... and more enum-classification comments assumed from here on :-)

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    // uart for the device
+    AP_HAL::UARTDriver *uart;
+    // shared scratch space for communication
+    uint8_t _buffer[RCDEVICE_PROTOCOL_MAX_PACKET_SIZE];
```
I don't think we need a shared scratch space.  We can afford 64 bytes on the stack - and the one place I really looked at required 4 IIUC!

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,319 @@
+    uint8_t _buffer[RCDEVICE_PROTOCOL_MAX_PACKET_SIZE];
+
+    // device is ready
+    bool _is_ready : 1;
```
As mentioned, some of these might go away with a suitable enum-class for State and a state variable.  We've found that generally bit-packing is a bad idea as it leads to increased code size ('though decreased RAM usage).

andyp1per replied:
Ok

peterbarker commented on libraries/AP_OSD/AP_OSD.cpp:
```diff
@@ -176,6 +179,7 @@ AP_OSD::AP_OSD()
     osd_type.set_default(HAL_OSD_TYPE_DEFAULT);
 #endif
     previous_pwm_screen = -1;
+    _singleton = this;
```
Need a sanity check for only-one-singleton.

andyp1per replied:
Ok

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -258,7 +264,14 @@ class AP_OSD
     void set_nav_info(NavInfo &nav_info);
-
+    // disable the display
+    void clear() {
```
Shouldn't this be `disable()`, then, with perhaps `clear()` as a side-effect?

andyp1per replied:
Ok

peterbarker commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -572,6 +574,25 @@ void RC_Channel::do_aux_function_camera_trigger(const aux_switch_pos_t ch_flag)
+            runcam->start_recording();
+            break;
+        default:
+            // low switch stops recording
```
Pointless (and insufficient as it doesn't mention it also handles `case MIDDLE`) comment.

andyp1per replied:
Ok

peterbarker commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -572,6 +574,25 @@ void RC_Channel::do_aux_function_camera_trigger(const aux_switch_pos_t ch_flag)
+
+    switch (ch_flag) {
+        case HIGH:
+            // high switch starts recording
```
Pointless comment.

andyp1per replied:
Ok

andyp1per commented on ArduCopter/Copter.cpp:
```diff
@@ -154,6 +154,8 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
 #endif
 #if CAMERA == ENABLED
     SCHED_TASK_CLASS(AP_Camera,            &copter.camera,              update_trigger,  50,  75),
+    SCHED_TASK_CLASS(AP_RunCam,            &copter.runcam,              update,          50,  50),
```
Ok

andyp1per commented on ArduCopter/Copter.h:
```diff
@@ -500,6 +501,7 @@ class Copter : public AP_Vehicle {
     // Camera
 #if CAMERA == ENABLED
     AP_Camera camera{MASK_LOG_CAMERA, current_loc};
+    AP_RunCam runcam{&rcmap};
```
Ok

andyp1per replied:
There really isn't an rcmap singleton - I've looked extensively!

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+const AP_Param::GroupInfo AP_RunCam::var_info[] = {
+    // @Param: FEATURES
+    // @DisplayName: RunCam features available
+    // @Description: The available features of the attached RunCam device. This should be determined automatically but can be set manually.
```
Ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+
+    // @Param: PROT_VER
+    // @DisplayName: RunCam protocol Version
+    // @Description: The protocol version supported by the attached RunCam device. Only protocol version 1 is supported.
```
Betaflight supports these first four parameters, so I presume they found some need for them. That said protocol seems particularly superfluous given that we only support 1 so I will take it out.

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+    // @User: Advanced
+    AP_GROUPINFO("PROT_VER", 1, AP_RunCam, _protocol_version, RCDEVICE_PROTOCOL_VERSION_1_0),
+
+    // @Param: INIT_NUM
```
BF config - I assume they found issues with some devices. I only have one to test on and it certainly takes some time to accept requests, but happy to bake in the defaults

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+    // @User: Advanced
+    AP_GROUPINFO("INIT_NUM", 2, AP_RunCam, _init_attempts, 6),
+
+    // @Param: INIT_INT
```
Not sure about this one. My experience and the overall impression I get is that the RunCam protocol is quite flakey depending on the device state. Get it into the wrong state or make requests too early and they will simply be dropped.

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+    // @User: Advanced
+    AP_GROUPINFO("INIT_INT", 3, AP_RunCam, _init_attempt_interval_ms, 1000),
+
+    // @Param: BT_DELAY
```
Ditto

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+    // @User: Advanced
+    AP_GROUPINFO("BT_DELAY", 4, AP_RunCam, _boot_delay_ms, 6000),
+
+    // @Param: BTN_DELAY
```
Ok we definitely need this. There is no way to tell when the camera has booted and is accepting requests other than waiting. Wait too short a time and everything gets out of sync, too long and its annoying. I am sure this will vary depending on camera version and power setup.

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+// singleton instance
+AP_RunCam *AP_RunCam::_singleton;
+
+AP_RunCam::RunCamResponse::RunCamExpectedResponseLength AP_RunCam::RunCamResponse::_expected_responses_length[RCDEVICE_NUM_EXPECTED_RESPONSES] = {
```
Ok!

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+// singleton instance
+AP_RunCam *AP_RunCam::_singleton;
+
+AP_RunCam::RunCamResponse::RunCamExpectedResponseLength AP_RunCam::RunCamResponse::_expected_responses_length[RCDEVICE_NUM_EXPECTED_RESPONSES] = {
```
Copied from BF, but I can change it!

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        uart = serial_manager->find_serial(AP_SerialManager::SerialProtocol_RunCam, 0);
+    }
+
+    if (uart != NULL) {
```
Ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+    }
+
+    if (uart != NULL) {
+        debug("initialized serial port\n");
```
Ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        debug("initialized serial port\n");
+        if (!_uart_started) {
+            uart->begin(115200);
+            _uart_started = true;
```
Copied from AP_BLHeli.cpp - maybe its not needed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+}
+
+// simulate pressing the camera button
+bool AP_RunCam::simulate_camera_button(RCCameraControlOperation operation)
```
Ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+    }
+
+    debug("press button %d\n", operation);
+    if (_protocol_version == RCDEVICE_PROTOCOL_VERSION_1_0) {
```
Makes sense

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        return;
+    }
+
+    if (is_initialized()) {
```
We want to record what comes from the RC switches - video_recording is a statement of intent. When we are fully initialized we will make the camera reflect the state. If we try and do this before the camera has booted it won't work.

andyp1per replied:
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+}
+
+// wait for the RunCam device to be fully ready
+bool AP_RunCam::is_initialized()
```
Ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        return;
+    }
+
+    if (is_initialized()) {
```
Ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+void AP_RunCam::update_osd()
+{
+    // only run the simulation process when configured, ready and not armed
+    if (!is_initialized() || (AP::arming().is_armed() && _in_menu == 0)) {
```
I've done some rough timings of how quickly you can press the switches. Video start/stop is especially slow. If _in_menu we will unwind the first time through and then won't come here again.

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+    receive();
+}
+
+// OSD update loop - runs at 5Hz
```
Ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+RC_Channel::aux_switch_pos_t AP_RunCam::get_channel_pos(uint8_t rcmapchan) const
+{
+    RC_Channel::aux_switch_pos_t position;
+    if (!rc().channel(rcmapchan-1)->read_3pos_switch(position)) {
```
Ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        return;
+    }
+
+    RC_Channel::aux_switch_pos_t throttle = get_channel_pos(_rcmapper->throttle());
```
ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        return RC_Channel::LOW;
+    }
+
+    return position;
```
Ok!

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+
+    // don't allow another command until the sticks have been re-centered
+    if (_is_button_pressed) {
+        if (yaw == RC_Channel::MIDDLE && pitch == RC_Channel::MIDDLE
```
Pretty sure the RCChannel code does, but I will check

peterbarker commented:
I take it we need to create an ObjectQeue for some reason, rather than just leaving them in the uart 'til we're good and ready?

WickedShell commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,774 @@
+    // @Description: The available features of the attached RunCam device. If 0 then the RunCam device will be queried for the features it supports, otherwise this setting is used.
+    // @User: Advanced
+    // @Bitmask: 0:Power Button,1:WiFi Button,2:Change Mode,3:5-Key OSD,4:Settings Access,5:DisplayPort,6:Start Recording,7:Stop Recording
+    AP_GROUPINFO("FEATURES", 0, AP_RunCam, _features, 0),
```
Don't use index 0 of any sub group like this, just bump all the indexes up by 1.

andyp1per replied:
Ok!

WickedShell commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,774 @@
+#include <AP_Logger/AP_Logger.h>
+#include <AP_OSD/AP_OSD.h>
+
+const AP_Param::GroupInfo AP_RunCam::var_info[] = {
```
Maybe add an enable param for this, so that people who aren't using runcams don't have the parameters coming down?

andyp1per replied:
I considered this, but nothing is enabled unless you special the runcam protocol on a serial device. If I add enable you have to both enable and select the protocol to get stuff working - which seems a little onerous, but happy to do this if that's the way we configure things. For comparison blheli protcol support does not have an enable flag.

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+
+// start the video
+void AP_RunCam::start_recording() {
+    if (_is_button_pressed) {
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+void AP_RunCam::update_2_key_simulation_process()
+{
+    if (AP::arming().is_armed()) {
+        // if for some reason the user has armed while in the menu
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+        if (_in_menu > 0) {
+            while (_in_menu-- > 0) {
+                simulate_camera_button(RCDEVICE_PROTOCOL_CAM_CTRL_CHANGE_MODE);
+                EXPECT_DELAY_MS(600);
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+// enter the 2 key OSD menu
+void AP_RunCam::enter_2_key_osd_menu()
+{
+    // can't change mode unless we are not recording
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+// run the 5-key OSD simulation process
+void AP_RunCam::update_5_key_simulation_process()
+{
+    if (AP::arming().is_armed()) {
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+}
+
+// handle a response
+void AP_RunCam::handle_5_key_simulation_response(RunCamResponse* response)
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+            debug("started serial port\n");
+        }
+
+        get_device_info();
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,831 @@
+{
+    flush_rx_buffer();
+
+    RunCamResponse* response = new RunCamResponse(this, commandID, param, timeout, maxRetryTimes, userInfo, parserFunc);
```
Fixed

andyp1per commented:
@peterbarker ok I think I have address all your (really great!) review comments. It works much better now with a state machine and the arming change is much more user friendly.

andyp1per commented:
@peterbarker here's another example where the MatekF405-Wing builds fine for me locally but fails in CI.

peterbarker commented:
On Sun, 8 Dec 2019, Andy Piper wrote:

> @peterbarker here's another example where the MatekF405-Wing builds fine for me locally but fails in CI.

OK, tridge is looking at killing smart batteries on these boards.  Stand 
by.

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+        }
+        break;
+
+    default:
```
Are not all events handled here?

Adding `default` means that if a new state is added it may get missed in handling somewhere.  If some state has no handling required, better to add an empty case for it.

This comment also applies to other `switch`es

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+    case Event::BUTTON_RELEASE:
+        send_5_key_OSD_cable_simulation_event(ev);
+        _waiting_device_response = true;
+        return;
```
return and break is overkill...

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+    if (response._result != ResponseStatus::SUCCESS) {
+        debug("response for command %d failed: %d\n", int(response._command), int(response._result));
+        simulation_OSD_cable_failed(response);
+        _waiting_device_response = false;
```
Move this to the top of the function, remove this copy and the one at the end of the method.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+        // the high 4 bits is the operationID that we sent
+        // the low 4 bits is the result code
+        _is_button_pressed = true;
+        ConnectionOperation operationID = ConnectionOperation(response._param);
```
`const`

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+        // the low 4 bits is the result code
+        _is_button_pressed = true;
+        ConnectionOperation operationID = ConnectionOperation(response._param);
+        uint8_t errorCode = (response._recv_buf[1] & 0x0F);
```
`const`

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+
+        // if data received done, trigger callback to parse response data, and update RUNCAM state
+        if (response._recv_response_length == response._expected_response_length) {
+            uint8_t crc = 0;
```
This could be a `crc()` method on the response object.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+            }
+        }
+        else {
+            _waiting_response_queue.update(response);
```
This is a lot of work here - everything in response will have to be copied into the ObjectBuffer object.  Probably better to keep one around and just push the final version into the ObjectBuffer.

andyp1per replied:
Ok, you win 🙂 I have removed the request queue.

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+            // response still in play
+            return true;
+        }
+        else if (response._max_retry_times > 0) {
```
pointless else

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+}
+
+// process all of the pending responses, retrying as necessary
+bool AP_RunCam::process_waiting(Response& response, uint32_t now)
```
`Response` is really `Request`.  It's a request that's waiting for a response.

andyp1per replied:
Gah! Ok you are right

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+            _waiting_response_queue.update(response);
+
+            return false;
+        } else {
```
pointless `else`  Please audit for more of these.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+    }
+
+    Response response;
+    while (process_waiting(response, AP_HAL::millis())) {
```
timeouts appear to be few and far between here.  Long-runninbg loops are bad on flight controllers, so we like to be over-cautious on out loops.  Explicit max-duration or limiting the number of bytes processed are typical.

peterbarker replied:
I'm finding the logic here rather difficult to follow.  Could you outline what you're trying to achieve here with this queue?

andyp1per replied:
Removed so fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+}
+
+
+// every time we send a packet to device and want to get a response
```
So why aren't we throwing away everything in the response queue?  Seems to be a little random as to what we're throwing away.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+        return;
+    }
+
+    uint8_t buffer[4];
```
I'd prefer to see this buffer assembled from its components statically rather than using the `*b++` stuff.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,350 @@
+        static Length _expected_responses_length[RUNCAM_NUM_EXPECTED_RESPONSES];
+    };
+
+    // ringbuffer of pending responses
```
Not really using it as a ringbuffer.

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_HAL/utility/RingBuffer.h:
```diff
@@ -146,14 +146,14 @@ class ObjectBuffer {
     
     /*
-      throw away an object
+      throw away an object from the front of the queue
```
Splitting this into a separate PR would be good.

andyp1per replied:
Fixed

peterbarker commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -572,6 +574,24 @@ void RC_Channel::do_aux_function_camera_trigger(const aux_switch_pos_t ch_flag)
+            runcam->start_recording();
+            break;
+        case LOW:
+        default:
```
Please avoid using `default` :-)

andyp1per replied:
Fixed

andyp1per commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -807,6 +808,32 @@ bool AP_Arming::fence_checks(bool display_failure)
+
+        // check camera is ready
+        const char *fail_msg = nullptr;
+        if (runcam->pre_arm_check(fail_msg)) {
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+
+// these are correct for the runcam split micro v2.4.4, others may vary
+// Video, Image, TV-OUT, Micro SD Card, General
+#define RUNCAM_TOP_MENU_LENGTH 6
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+    // only run the simulation process when configured, ready and not armed
+    if (AP::arming().is_armed()) {
+        update_state_machine_armed();
+    }
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+// OSD update loop
+void AP_RunCam::update_osd()
+{
+    // only run the simulation process when configured, ready and not armed
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+// return radio values as LOW, MIDDLE, HIGH
+RC_Channel::aux_switch_pos_t AP_RunCam::get_channel_pos(uint8_t rcmapchan) const
+{
+    RC_Channel::aux_switch_pos_t position = RC_Channel::LOW;
```
It can't be.

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+RC_Channel::aux_switch_pos_t AP_RunCam::get_channel_pos(uint8_t rcmapchan) const
+{
+    RC_Channel::aux_switch_pos_t position = RC_Channel::LOW;
+    RC_Channel* chan = rc().channel(rcmapchan-1);
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+// update the state machine when armed or flying
+void AP_RunCam::update_state_machine_armed()
+{
+    uint32_t now = AP_HAL::millis();
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+        return;
+    }
+
+    _transition_start_ms = now;
```
It's needed.

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+}
+
+// update the state machine when disarmed
+void AP_RunCam::update_state_machine()
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+// update the state machine when disarmed
+void AP_RunCam::update_state_machine()
+{
+    uint32_t now = AP_HAL::millis();
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+    _transition_start_ms = now;
+    _transition_timeout_ms = 0;
+
+    Event ev = map_rc_input_to_event();
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+
+    Event ev = map_rc_input_to_event();
+    if (ev == Event::BUTTON_RELEASE) {
+        _is_button_pressed = false;
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+        simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_CHANGE_STOP_RECORDING);
+        set_mode_change_timeout();
+        _state = State::READY;
+    }
```
Fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,892 @@
+{
+    if (has_feature(Feature::RCDEVICE_PROTOCOL_FEATURE_SIMULATE_5_KEY_OSD_CABLE)) {
+        handle_5_key_simulation_process(ev);
+    }
```
Fixed all

andyp1per commented on APMrover2/Rover.cpp:
```diff
@@ -82,6 +82,10 @@ const AP_Scheduler::Task Rover::scheduler_tasks[] = {
+#endif
+#if RUNCAM_ENABLED == ENABLED
+    SCHED_TASK_CLASS(AP_RunCam,            &rover.runcam,         update,          50,  50),
+    SCHED_TASK_CLASS(AP_RunCam,            &rover.runcam,         update_osd,      5,   200),
```
I went back and forth on this. In the end I think you are right.

andyp1per commented:
@peterbarker so I think I have addressed all your comments. It's not quite working correctly at the moment with timeouts, it's not far off but I am out of time for debugging for a few days. You can see the shape so any further review appreciated.

andyp1per commented:
@peterbarker all working now, rebased and fresh!

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -258,7 +264,14 @@ class AP_OSD
+        _disable = true;
+    }
+    // enable the display
+    void display() {
```
This should probably be `enable`

andyp1per replied:
ok

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,919 @@
+    { Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION, 3 },
+};
+
+// these are correct for the runcam split micro v2.4.4, others may vary
```
What happens if the menus are of different lengths?

Is it firmware-version dependant?

andyp1per replied:
I don't know but a description of a menu I found for older firmware was different, so probably. Ultimately this should be configurable, but not quite sure how

peterbarker commented on libraries/AP_Camera/AP_RunCam.cpp:
```diff
@@ -0,0 +1,919 @@
+        return;
+    }
+
+    debug("found serial port\n");
```
Can probably kill this one.

andyp1per replied:
ok

andyp1per commented on APMrover2/Rover.h:
```diff
@@ -192,6 +196,9 @@ class Rover : public AP_Vehicle {
     AP_Camera camera{MASK_LOG_CAMERA, current_loc};
 #endif
+#if RUNCAM_ENABLED == ENABLED
+    AP_RunCam runcam;
```
It's a good suggestion. In general how should stuff be switched off - right now there is no consistency, no-where where you can do it in a single place? By default we assume everything is on. It feels like waf should look at a central config location and allow command-line deselection of things.

andyp1per replied:
Moved to AP_Vehicle

andyp1per commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,358 @@
+#include <AP_OSD/AP_OSD.h>
+
+// the protocol for Runcam Device definition
+#define RUNCAM_HEADER 0xCC
```
ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,358 @@
+        return _singleton;
+    }
+
+    // definitions prefiexed with RCDEVICE taken from https://support.runcam.com/hc/en-us/articles/360014537794-RunCam-Device-Protocol
```
Some of the names get ambiguous and we don't do that anywhere else for other protocols that we can see, so I would prefer to leave as is

andyp1per commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,358 @@
+    // Check whether arming is allowed
+    bool pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) const;
+
+    static const uint32_t RUNCAM_MAX_PACKET_SIZE =       64;
```
ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,358 @@
+    // camera state
+    State _state = State::INITIALIZING;
+    // count of ticks since last OSD cycle
+    uint32_t _tick_count;
```
So I did it this way because decimation was easy and the rate of the OSD doesn't have to be super well controlled. If we start having to write a partial scheduler inside update() then I am against doing it this way - it's just duplicating somewhat tricky code code little obvious benefit.

andyp1per replied:
Ok fixed

andyp1per commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,358 @@
+    // shared inbound scratch space
+    uint8_t _recv_buf[RUNCAM_MAX_PACKET_SIZE]; // all the response contexts use same recv buffer
+
+    class Request;
```
It's used inside Request and C++ didn't like the circular reference

andyp1per commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,358 @@
+        RequestStatus _result; // whether we were successful or not
+
+        // get the length of the expected response
+        uint8_t get_response_length(const Command command);
```
ok

andyp1per commented on libraries/AP_Camera/AP_RunCam.h:
```diff
@@ -0,0 +1,358 @@
+        }
+
+        struct Length
+        {
```
ok

peterbarker commented:
On Thu, 12 Dec 2019, Andy Piper wrote:

> It's a good suggestion. In general how should stuff be switched off - right now there is no consistency, no-where where you can do it in a single place?

Correct.  Ish.  On ChibiOS you can put things in the hwdef files.

> By default we assume everything is on. It feels like waf should look at a central config location and allow command-line deselection of things.

We've often mused on just that.  Sadly, we've never gotten a PR to do it.

One issue is that at configure you don't know which vehicle you're 
building so you don't know which options are relevant.  One solution there 
would be to present the union of all vehicle options, so you might have 
"--no-copter-nav-guided" and "--no-plane-soaring"  Note the namespacing 
requirements in there, too.

Another note is that those plane options shouldn't be visible to the 
libraries (in much the same way that the existing plane options aren't). 
We need another set of global-only  options (e.g. a global STATS_ENABLED 
equivalent)

Further, you need to have different defaults for the libraries based on 
which vehicle you are compiling for - e.g. Copter and Rover want proximity 
enabled by default, Plane doesn't.  Or, preferably, allow the vehicle to 
override the default with its own default....

I actually poked at this in mid-2018.  I didn't get very far: 
https://github.com/ArduPilot/ardupilot/compare/master...peterbarker:wip/config-libs?expand=1

peterbarker commented:
On Thu, 12 Dec 2019, Andy Piper wrote:

> So I did it this way because decimation was easy and the rate of the OSD doesn't have to be super well controlled. If we start having to write a partial
> scheduler inside update() then I am against doing it this way - it's just duplicating somewhat tricky code code little obvious benefit.

It's not an uncommon pattern, particularly for logging.

https://github.com/ardupilot/ardupilot/blob/master/libraries/GCS_MAVLink/GCS_Common.cpp#L1376
https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_AHRS/AP_AHRS_NavEKF.cpp#L346
https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_Notify/OreoLED_I2C.cpp#L447

We've had some subtle bugs due to different scheduler rates in the past.

andyp1per commented:
@peterbarker this is overflowing on the matek 405 wing again by 44 bytes. Seems like this is going to be an ongoing problem. I've disabled on copter for now.

andyp1per commented:
Need CAM_TYPE, CAM_CONTROL and move to AP_Vehicle

andyp1per commented:
@tridge last 4 commits generic vehicle management as discussed but for copter only. Will do the others once we have got copter right.

manotroll commented:
can i test on drone case passes and cheap px4 compatible firmware

andyp1per commented:
Rebased

tridge commented:
@andyp1per as we discussed, I've changed CAM_RC_TYPE to be an enable parameter, but made it auto-enable if you have the SERIAL protocol setup.
I've happy for this PR to be merged once it passes CI. Many thanks for you efforts on this!
I know that the runcam support isn't yet proven to work for everyone, but that can be worked on after we have it in master. I want to merge this early so we can start to use the new AP_Vehicle parameter and scheduling components

rmackay9 commented:
I think we've introduced a small issue here by adding a 17th bit to the ArmingChecks enum without increasing the get_enabled_checks() return value (a uint16_t) and we also need to increase the size of the log_Arm_Disarm log structure.

andyp1per commented:
@rmackay9 good catch!

aaaaaaudrey commented:
My apologies if this should be in a separate issue:

It seems that there are some strange issues with this implementation. 

The runcam menu ignores RC reversals in ardupilot. in my case, pitch stick up moved the menu selection down. Reversing in ardupilot did not work. I had to use radio reversal to fix the runcam menu and then undo that reversal (to make my sticks work properly) in ardupilot. 

Rolling hard-right does not always enter the camera menu, while yawing hard right always does. In some cases, seemingly randomly, I get the "remote mode" popup from the camera after using roll to enter the menu, but the only way to get to the actual camera menu is to use yaw. 

Sometimes, repeatedly pushing roll hard left makes ardupilot think that it is no longer connected to the camera, while the camera menu is still definitely active, and thus the outgoing connection to the camera is broken, making it impossible to close the menu without power cycling or unplugging the camera. I can provide video examples of all of these issues if that would be helpful.

andyp1per commented:
Hi @aaaaaaudrey can you raise a github issue for this?

Please can you describe:
- Your settings / parameters
- Which runcam camera you are using
- What sequence of commands from boot causes the problem (it's a state machine so I need the exact sequence of commands from the time you boot the board to reproduce the issue).

It looks like you are using a bare camera with the remote protocol. I'm afraid this got less testing than the split version and is even more timing sensitive because my copters all have splits on them.

aaaaaaudrey commented:
@andyp1per #14540

wcfung commented:
Not sure if there is anyone still following this thread.

 I am trying to conenct a Runcam Split (1st version) to Pixhawk (running Arduplane v4.0.7) in order to start/stop video. Have already connected the Split module Gnd, Rx Tx to Pixhawk UART Gnd, Tx, Rx respectively.  Mission Planner (1.3.74) Serialn_PROTOCOL is also set to 26 (for Runcam). I am not sure why 26 is not an option in the parameter list. Serialn_BAUD is left at 57600. 

After power up, the Runcam Split is not recognized in the initialization (shown in message tab) as indicated in ardupilot wiki. I tried reversing the signal wires as suggested in wiki. This did not change anything. The camera should be in UART control mode, which is usually the default mode for Split camera. However I cannot confirm this as I had difficulty connecting the camera to Runcam App. The camera recording function is normal. 

Pls suggest any possible causes that prevent Pixhawk from recognising the Runcam Split.

andyp1per commented:
@wcfung this feature is only supported in Ardupilot 4.1 which is now out in beta

wcfung commented:
Thank you for yr reply. I am glad you are still following this thread. But you did mention on Nov 16, 2019 in this thread that "video start/stop is now working".  In 2019, it was before Ardupilot 4.0.7. Did I miss something ?

Appreciate your help.

andyp1per commented:
It would have had to have been before 4.0.0 to have made it into 4.0, anything after that is bugfixes only

wcfung commented:
I don't understand the first part of the sentence. Anyway, my understanding
is when it is V4.1 official release(for Copter only),
then the start/stop video function will work.

However, I will use this function in Arduplane, which is not in V4.1 beta
phase. Seems I have to wait a bit longer.

Anyway, thanks.

On Wed, Apr 21, 2021 at 11:01 AM Andy Piper ***@***.***>
wrote:

> It would have had to have been before 4.0.0 to have made it into 4.0,
> anything after that is bugfixes only
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/12810#issuecomment-824251745>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD5VK752EBPTSSJWCTUTKUTTJ4HHDANCNFSM4JLOEAXQ>
> .
>
